### ISSUE TITLE
Port twisted.protocols.basic.FileSender to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6026]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6026|
|Type|enhancement|
|Created|2012-09-28 16:56:21Z|
|Branch|https://github.com/twisted/twisted/tree/filesender-py3-6026-2|

`FileSender` is not being ported to Python 3 as part of [#6025](https://github.com/twisted/twisted/issues/6025), so will need to be done separately.

<details><summary>Searchable metadata</summary>

```
trac-id__6026 6026
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_filesender_py3_6026_2 branches/filesender-py3-6026-2
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1348851381000000 1348851381000000
changetime__1360440137000000 1360440137000000
version__None None
owner__tom_prince tom.prince

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [36403]) Branching to 'filesender-py3-6026'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

This is ready for review. This was mainly about adding new tests, as there was no incompatible bits that I could spot. I did change the interface implemented and remove one method not part of the interface, though.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_4"></a>|@tomprince commented|
|-|-|

[buildbot run](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/filesender-py3-6025)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

I clicked on the given link but I didn't actually see build results.  Is there a buildbot bug here someplace, or is 4 days actually so long that our build history falls off the end?

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince commented|
|-|-|

Typo in the link: [here](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/filesender-py3-6026)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_7"></a>|@tomprince set owner to @therve|
|-|-|

  1. I don't think removing `pauseProducing` without deprecation is allowable, under the compatibility policy.
  1. Is there any reason to use a real file, rather than something like `StringIO`? Also, the consumer should probably be something designed for testing like `StringTransport`.
  1. New tests should be synchronous when possible, using `successResultOf`, `failureResultOf`, and `assertNoResult`.
  1. The tests don't test the case when multiple chunks are required to send the file. (It is probably reasonable to shrink `CHUNK_SIZE` when testing this.

Please resubmit for review, after addressing the above points.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [37088]) Branching to 'filesender-py3-6026-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

Thanks, everything handled I think!

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_11"></a>|@tomprince commented|
|-|-|

  1. The comments in the test should explain why `resumeProducing` needs to be called twice.
  1. `test_transferMultipleChunks` doesn't actually test that things are sent in chunk-size bits. It should check the output at each step.
  1. Something should verify that the producer gets unregistered.
  1. Similarly, something should verify that the producer gets properly registered as a non-streaming producer.

I'll fix these up and commit.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_12"></a>|@tomprince set owner to @tomprince|
|-|-|

Also, `stopProducing` isn't normally called, except in exception cases, so don't call it from the tests except in exceptional cases.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince set status to `closed`|
|-|-|

(In [37101]) Merge filesender-py3-6026-2: Port twisted.protocols.basic.FileSender to Python 3.

Author: therve
Reviewers: tom.prince
Fixes: [#6026](https://github.com/twisted/twisted/issues/6026)

The interface implemented changed to the more specific IPullProducer.
